### PR TITLE
Static routes monitoring metrics added to network dashboard BP

### PR DESCRIPTION
**Pull request by maunope**
Added static routes metrics, for single VPC and for PPG, a few other minor refactorings

### CODE REVIEW

**Comment by juliocc** [comment]
Could you please add a description and/or a descriptive title?

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 1]
```diff
@@ -14,9 +14,14 @@
 # limitations under the License.
 #
 
+from ast import AnnAssign
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:17]: Do we need this import?

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 2]
```diff
@@ -14,9 +14,14 @@
...
+from os import times
+import re
+from sqlite3 import Timestamp
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:20]: Do we need this import?

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 3]
```diff
@@ -169,11 +175,121 @@ def get_dynamic_routes_ppg(config, metric_dict, usage_dict, limit_dict):
...
-                                   limit_dict)
+                                   limit_dict, timestamp)
+      print(
+          f"Buffered {metric_dict['usage']['name']} for peering group {network_dict['network_name']} in {project_id}"
+      )
+
+
+def get_static_routes_dict(config):
+  '''
+    Calls the Asset Inventory API to get all static customr routes under the GCP organization.
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:190]: typo: customr -> custom

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 4]
```diff
@@ -169,11 +175,121 @@ def get_dynamic_routes_ppg(config, metric_dict, usage_dict, limit_dict):
...
+          "asset_types": ["compute.googleapis.com/Route"],
+          "read_mask": read_mask
+      })
+
+  for resource in response:
+    for versioned in resource.versioned_resources:
+      static_route = dict()
+      for field_name, field_value in versioned.resource.items():
+        static_route[field_name] = field_value
+      static_route["project_id"] = re.search("\/([^\/]*)(\/[^\/]*){3}$",
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:214]: I prefer using splits like field_value.split('/')[-1] to get the network name and field_value.split('/')[6] (to be checked) to get the project_id, I think it is more readable then regexp. What do you think?

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 5]
```diff
@@ -169,11 +175,121 @@ def get_dynamic_routes_ppg(config, metric_dict, usage_dict, limit_dict):
...
+      Returns:
+        None
+  '''
+  timestamp = time.time()
+  project_usage = {project: 0 for project in config["monitored_projects"]}
+
+  #usage is drilled down by network
+  for network_link in static_routes_dict:
+
+    project_id = re.search("\/([^\/]*)(\/[^\/]*){3}$", network_link).group(1)
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:260]: same about the split vs regexp

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 6]
```diff
@@ -169,11 +175,121 @@ def get_dynamic_routes_ppg(config, metric_dict, usage_dict, limit_dict):
...
+    if (project_id not in config["monitored_projects"]):
+      continue
+    network_name = re.search("\/([^\/]*)$", network_link).group(1)
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:263]: same about split vs regexp

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 7]
```diff
@@ -169,11 +175,121 @@ def get_dynamic_routes_ppg(config, metric_dict, usage_dict, limit_dict):
...
+    project_usage[project_id] = project_usage[project_id] + static_routes_dict[
+        network_link]
+
+    metric_labels = {"project": project_id, "network_name": network_name}
+    metrics.append_data_to_series_buffer(
+        config, metrics_dict["metrics_per_network"]["static_routes_per_network"]
+        ["usage"]["name"], static_routes_dict[network_link], metric_labels,
+        timestamp=timestamp)
+
+  #limit and utilization are calculated by projec
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:274]: typo projec -> project

**Review by aurelienlegrand — CHANGES_REQUESTED** [review]
Thanks Maurizio for another contribution. 

Just to clarify, we have different limits:
- static routes per VPC peering group
- static routes per project (regional + global if global routing mode)

You are implementing both, right? (the global part is fine for now, the regional part can come later)

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 1 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:17]: removed

Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:17]: removed

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 8]
```diff
@@ -14,9 +14,14 @@
 # limitations under the License.
 #
 
+from ast import AnnAssign
+from os import times
```
[the remark this answers is not in the document]
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:18]: removed

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 2 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:20]: removed

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 3 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:190]: k

**Comment by maunope** [comment]
> Thanks Maurizio for another contribution.
> 
> Just to clarify, we have different limits:
> 
> * static routes per VPC peering group
> * static routes per project (regional + global if global routing mode)
> 
> You are implementing both, right? (the global part is fine for now, the regional part can come later)

Both, for global only. added an assumption in readme.md explaining it only works with global for now, and with peering routes always exchanged

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 7 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:274]: k

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 4 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:214]: k

Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:214]: k

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 5 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:260]: k

File: blueprints/cloud-operations/network-dashboard/README.md [hunk 9]
```diff
@@ -46,22 +46,25 @@ The Cloud Function currently tracks usage, limit and utilization of:
...
+- Static routes per project (VPC drill down is available for usage)
+- Static routes per VPC peering group (note: assumes custom routes sharing is ON for all peered networks)
 - IP utilization per subnet (% of IP addresses used in a subnet)
 - VPC firewall rules per project (VPC drill down is available for usage)
 - Tuples per Firewall Policy
 
 It writes this values to custom metrics in Cloud Monitoring and creates a dashboard to visualize the current utilization of these metrics in Cloud Monitoring.
 
-Note that metrics are created in the cloud-function/metrics.yaml file.
+Note that metrics are created in the cloud-function/metrics.yaml file. also note that the Cloud Function assumes all VPCs in peering groups are within the same organization.
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/README.md:59]: Why this remark about the same organization?

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py [hunk 10]
```diff
@@ -14,9 +14,11 @@
 # limitations under the License.
 #
 
+import re
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:17]: I think this import is not required anymore.

**Review by aurelienlegrand — APPROVED** [review]
LGTM

File: blueprints/cloud-operations/network-dashboard/README.md (hunk 9 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/README.md:59]: On paper, it is possible to peer VPCs from other orgs, now that I think of it, it should work fine if permissions are right. how about we peer a VPC between our test orgs and see it working e2e? then once we're sure we can change this to something like "it is assumed the CF service account is granted permissions in external orgs, those permissions are not set by the TF in this repo"

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py (hunk 10 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/routes.py:17]: k

**Review by aurelienlegrand — APPROVED** [review]
LGTM

**Pull request merged**